p2p/src/lib.rs

```diff
@@ -117,24 +184,90 @@ where
-                .await
-                .map(|socket| self.create_peer(socket, PeerRole::Outbound))?,
+            ConnectivityEvent::Accept(peer_id, socket) => {
+                self.create_peer(peer_id, socket, PeerRole::Inbound)
+            }
+            ConnectivityEvent::Connect(address) => {
+                self.network.connect(address).await.map(|(peer_id, socket)| {
+                    self.create_peer(peer_id, socket, PeerRole::Outbound)
+                })?
+            }
         }
 
         Ok(())
     }
 
-    fn peer_discovered(&mut self, peers: &[NetworkingBackend::Address]) -> error::Result<()> {
-        println!("peers discovered: {:#?}", peers);
+    /// Try to establish new outbound connections if the total number of
+    /// active connections the local node has is below threshold
+    ///
+    // TODO: move all peer management to separate file
+    async fn auto_connect(&mut self) -> error::Result<()> {
+        // we have enough active connections
+        if self.peers.len() >= MAX_ACTIVE_CONNECTIONS {
+            return Ok(());
+        }
+
+        // we don't know of any peers
+        if self.discovered.is_empty() {
+            return Err(P2pError::NoPeers);
+        }
+
+        let npeers = std::cmp::min(
+            self.discovered.len(),
+            MAX_ACTIVE_CONNECTIONS - self.peers.len(),
+        );
+
+        // TODO: improve peer selection
+        let mut peers = Vec::new();
+        let mut iter = self.discovered.iter();
+
+        for _ in 0..npeers {
+            let peer_info = iter.next().expect("Peer to exist");
+            let (ip4, ip6) = match peer_info.1 {
+                PeerAddrInfo::Raw { ip4, ip6 } => (ip4, ip6),
+            };
+            assert!(!ip4.is_empty() || !ip6.is_empty());
+
+            // TODO: let user specify their preference?
+            let addr = if ip6.is_empty() {
+                Arc::clone(ip4.iter().next().unwrap())
+            } else {
+                Arc::clone(ip6.iter().next().unwrap())
+            };
+
+            peers.push((*peer_info.0, addr));
+        }
+
+        for (id, addr) in peers.into_iter() {
+            self.discovered.remove(&id);
+            let _ = self.on_connectivity_event(ConnectivityEvent::Connect((*addr).clone())).await;
+        }
+
         Ok(())
     }
 
-    fn peer_expired(&mut self, peers: &[NetworkingBackend::Address]) -> error::Result<()> {
-        println!("peers expired: {:#?}", peers);
+    /// Update the list of peers we know about or update a known peers list of addresses
+    fn peer_discovered(&mut self, peers: &[net::AddrInfo<NetworkingBackend>]) -> error::Result<()> {
+        for info in peers.iter() {
+            if self.peers.contains_key(&info.id) {
+                continue;
+            }
+
+            match self.discovered.entry(info.id).or_insert_with(|| PeerAddrInfo::Raw {
+                ip4: HashSet::new(),
+                ip6: HashSet::new(),
+            }) {
```

Comment:
Maybe there's something wrong with my IDE; though please format the file.

Reply:
It looks really ugly but the code is formatted. I believe if we used nightly, we'd have access to `PeerAddrInfo::Raw::default()` which would be much cleaner. But that alone probably doesn't warrant the switch